lib/http/2/compressor.rb

```diff
@@ -553,14 +553,15 @@ def decode(buf)
         decoding_pseudo_headers = true
         until buf.empty?
           next_header = @cc.process(header(buf))
+          next if next_header.nil?
```

Comment:
> Why would it be nil?

it seems that nginx's (1.14 and 1.15, at least) http2 response may include content (a space?) that causes the decompressor hit a `nil` within the new header-order checking code when attempting to send `#first` to `next_header`.

> Do we have a failure test case?

i've attempted to add one that illustrates the problem, but request help in placing the test correctly within the specs.

> On a related note, this wasn't an issue prior to recent 0.10.0 release?

if you try this [spec](https://github.com/kenichi/http-2/blob/header_nils/spec/compressor_spec.rb#L598-L608) on `v0.9.1`, it passes; `v0.10.0` it fails, causing clients using that version of the library to fail when talking to nginx.

i've pushed a basic testing support [Dockerfile](https://github.com/kenichi/h2/blob/master/test/support/nginx/Dockerfile) to my client's repo. this uses stable versions of nginx and openssl. you can try it by using the rake task `test:nginx`. once that is running, from this repo, i've tested it with the following:

```
$ bundle exec ruby example/client.rb https://localhost:4430/nginx.conf
```

| http-2 version | response |
|---|---|
| 0.9.1 | 200 |
| 0.10.0 | HTTP2::Error::InternalError exception: undefined method `first' for nil:NilClass - closing socket |
| 0.10.0-header_nils | 200 |

Reply:
I was thinking the same thing... but was wondering why emit was initialized to nil to begin with?

Reply:
Looking back, most likely an oversight. We need to declare it at the start to ensure that the scoping + return works as expected.

Reply:
I lean towards avoiding object allocation and additional iterations like `#compact`.

If we go with option 2, then in order to avoid NMEs on nils, `Decompressor#decode` would either need to check for `next_header` being an empty array, or `.first` returning `nil` here:

https://github.com/igrigorik/http-2/blob/master/lib/http/2/compressor.rb#L556

Then, if I understand correctly, this `list` is returned all the way up to `:promise_headers` and `:headers` Stream event handlers. Therefore, the check mentioned above would also need to keep empty arrays from being pushed on to `list`, or another compact-like iteration would have to be made to remove them. 

So my opinion is to go with option 1 and update the docs for `#process`.

Reply:
Yes, that's what I meant with, "check for next_header being an empty array," and if you did that, you wouldn't need the check on the .first ("or").

But, if you're going to check an array being empty, why not check a variable being nil instead? Same #LOC and one less Array alloc _per stream_.

Reply:
Fair enough. I'm good with the code patch you're proposing, only blocker is cleaning up the test so we know what we're testing. :)